### DISCUSSION TITLE
Xbox Series controller Share button

### DISCUSSION THREAD

**Discussion by ld892012** [AntiMicroX/antimicrox#797] [Q&A] [answered]
The share button seems to be recognized by "Button 12" in controller mappings but i have to replace another button by it to make it assignable.

Is there no way to add the share button of Xbox series controller to the main view of AntiMicroX without lose any buttons ?

**Comment by ld892012** [comment, marked as answer]
I find two workaround to make it work, first time you need to get your controller GUID :
- Download and open [SDL2 Gamepad Mapper](https://gitlab.com/ryochan7/sdl2-gamepad-mapper/-/releases)
- Plug Xbox Series controller
- Click on "Copy Gamepad GUID"

Then, I do the following :

**Method 1 (Use Controller Mappings):**
We can't just create a mapping with AntimicroX because it use UniqueID, but there is a way to change UniqueId to Controller GUID by editing AntimicroX's config file
- Paste your Gamepad GUID somewhere, you will have something like this : `03XXXXXXYYYYYYYYYYYYYYYYYYYYYYYY`
- Just replace each `X` to `0` and you have your Controller GUID
- Open AntimicroX
- Click on "Controller Mapping" (remember UniqueId if you already used Controller Mappings) then click on "Save"
- Close AntimicroX
- Open `antimicrox_settings.ini` in `%LocalAppData%\AntimicroX` (installed version) or `<AntimicroX folder>\bin` (portable version)
- Go to `[Mappings]` section, you will see something like this :
```
<UniqueId>="<UniqueId>,<Controller Name>,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,"
<UniqueId>Disable=0
```
- Now replace `<UniqueId>` with your `<Controller GUID>` and add `touchpad:b11,` at the end of line (before the `"`)
- If you want, you can change `<Controller Name>` with anything you want (this can be useful to see more easily if the modification is used)
- Start AntimicroX, now Capture button as mapped to button 21

If this method don't work, it's maybe because Controller GUID was badly edited, anyway you can try second method

**Method 2 (modify GameControllerDB) :**
- Open `gamecontrollerdb.txt` located into `<AntiMicroX folder>\share\antimicrox`
- Add a line with the following :
```
<Gamepad GUID>,Xbox Series Controller,touchpad:b11,a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b10,leftshoulder:b4,leftstick:b8,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b9,righttrigger:a5,rightx:a2,righty:a3,start:b7,x:b2,y:b3,platform:Windows,
```

- Save the file and open AntimicroX, now Capture button as mapped to button 21

But this file is overwritten by each AntimicroX update, so we have to add this line after each updates

Prefer use Method 1 over Method 2 because your changes will not be erased after an update

**Reply by realgooseman** [reply]
Thank you for this @ld892012,

Method 1 was unclear for me so I tried Method 2 and it works perfectly. The Share button is finally detected (as button 21).
Do we have to repeat this after every AntiMicroX update ? Is there a set and forget method to simplify things ?

**Reply by ld892012** [reply]
Yes because `gamecontrollerdb.txt` is overwritten after an update, but now Method 1 is better because i have find a way to not use UniqueId for "Controller Mappings" feature
So i edited my answer to update my first method for add this and i makes it more understandable, you can try first method if you want, it's a "set and forget" method unlike Method 2

**Reply by realgooseman** [reply]
Thanks for updating Method 1 but:


1. I don't have a `[Mappings]` section in `antimicrox_settings.ini`. Should I create that section manually then ?

> _Sections I got: General, DefaultAutoProfileAll, AutoProfiles, Notifications, Mouse and Controllers._


2. The data under  `[Mappings]`  section should look something like this right ?


```
<UniqueId>="03XXXXXXXXXXXXX000000007200,Xbox Series Controller,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,touchpad:b11,"
<UniqueId>Disable=0
```

**Reply by realgooseman** [reply]
Just created the `[Mappings]` section and added the lines above and it works great, so now it should stay like that even after updates as it's in the `antimicrox_settings.ini` user settings and not in the `gamecontrollerdb.txt` database right ?

Thanks a lot for the help @ld892012 !

**Reply by ld892012** [reply]
> I don't have a `[Mappings]` section in `antimicrox_settings.ini`. Should I create that section manually then ?

You can create it manually or you can use AntimicroX by click on "Controller Mapping" then click on "Save", the section is automatically create with the mapping of the controller, after that you just have to add capture button mapping and if you have several Xbox Series controllers you can replace `UniqueId` used by AntimicroX by `Controller GUID` (it avoid you to do this method for each Xbox series controllers you have)



> ```
> <UniqueId>="03XXXXXXXXXXXXX000000007200,Xbox Series Controller,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,touchpad:b11,"
> <UniqueId>Disable=0
> ```

It seems you have missing numbers, GUID have 32 length, but your GUID contains only 28, just keep your first 2 chars and replace next 6 by 6 zeroes
For example `030012345e040000ff02000000007200` gives `030000005e040000ff02000000007200`
And you can replace `<UniqueId>` too


> so now it should stay like that even after updates as it's in the `antimicrox_settings.ini` user settings and not in the `gamecontrollerdb.txt` database right ?

Yes because settings are not erased after an update

Glad to help !

**Reply by realgooseman** [reply]
Yes, I figured that out thanks.



> "And you can replace `<UniqueId>` too"

By that, do you mean that the `<UniqueId>` prefix can be replaced with the Controller GUID as well ?
So for me, something like this:


```
0300509d5e040000120b000000007200="0300509d5e040000120b000000007200,Xbox Series Controller Custom,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,touchpad:b11,"
0300509d5e040000120b000000007200Disable=0
```

**Reply by ld892012** [reply]
Yes but AntimicroX don't use it if the GUID not start with `03000000` so you can replace all your `0300509d5e040000120b000000007200` by `030000005e040000120b000000007200` to give that :
```
030000005e040000120b000000007200="030000005e040000120b000000007200,Xbox Series Controller Custom,platform:Windows,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,touchpad:b11,"
030000005e040000120b000000007200Disable=0
```

I have to do this in my side because AntimicroX don't apply my mapping if i don't replace that chars

**Reply by XxiDonxX** [reply]
can you help me out with this? I am trying to map my share button to a specific key but the share button doesn't appear on antimicrox, how can I add it or make it appear?

**Reply by ld892012** [reply]
Normally with all information above you can make share button work.
Where are you stuck ?

**Reply by SixelAlexiS90** [reply]
@ld892012 
Hi! I've followed your guide but, at least for Antimicro X 3.4.1, it's a bit different.
Specifically:

> Click on "Controller Mapping" (remember UniqueId if you already used Controller Mappings) then click on "Save"

There is no "Save" button in "Options > Settings > Controller Mappings", you can only "Add" or "Delete" the GUID.

So I've put the "Controller GUID" in it, clicked "OK" and closed the program.
At that point in `antimicrox_settings.ini` there was this on the bottom:

> [Mappings]
03000000YYYYYYYYYYYYYYYYYYYYYYYYDisable=0

Then I've followed your template while using my "Controller GUID" and it worked, the "Share button" was bind on Button 21.
By the way I've tried using the "Gamepad GUID" directly with the same template and works the same, is there a specific reason to use the "Controller GUID" instead?

Thank you very much in any case but yeah, I would tweak the guide to make it even more clear and fix those differences with the latest build.

p.s. I think that the **<>** can be confusing when specifying a text to replace, since at first I was leaving them when following the template.

**Reply by websy-subtle** [reply]
In my experience, you have to set the config file as "read-only", because AMX would change the GUID value at some point. Thanks for the solution!

PD: Changing the numbers from the third to the eighth position to zeros seems to be only for wireless connections. If you're setting up a wired connection, the GUID is different, but also leave it just like it is.

**Reply by Buzbee** [reply]
I had to use the whole 40 characters as Gamepad GUID and it finally worked for me.
I just clicked on "Controller Mapping" in the bottom left corner of AntiMicroX and copied everything in the box under "SDL 2 Game Controller Mapping" by highlighting it and scrolling all the up and then added "touchpad:b11" after moving "platform:Windows," to the end. I named it Xbox Series S/X Controller to confirm it was pulling it from the gamecontrollerdb.txt. I pasted it on top of the file right under "# Windows". Now I can use the Share button with AntiMicroX by assigning Button 21 to a key. I also had to make the gamecontrollerdb.txt read-only or else it would be deleted on close. Hope this helps someone.

Copied from "SDL 2 Game Controller Mapping":
030079c35e040000130b00000905720011182835,Xbox Series **X** Controller,**platform:Windows**,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,


Pasted into "gamecontrollerdb.txt":
030079c35e040000130b00000905720011182835,Xbox Series **S/X** Controller,**touchpad:b11**,a:b0,b:b1,x:b2,y:b3,back:b6,start:b7,guide:b10,leftshoulder:b4,rightshoulder:b5,leftstick:b8,rightstick:b9,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,dpup:h0.1,dpleft:h0.8,dpdown:h0.4,dpright:h0.2,**platform:Windows,**
